Product List: Replace full product objects with simplified objects

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

Part of WOOMOB-619

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR continues the work in #15959 to improve the performance on the product list by loading simplified product objects instead of full ones.

Changes include:
- Update `ProductListItem` with new stock-related properties needed for displaying on product list items.
- Update `selectFirstProductIfAvailable` in product list to fetch full object for only the item at first index, not the whole list.
- Update product list datasource to work with simplified objects.
- Update the product list cell to accept simplified objects.

Some helpers added:
- Add `toListItem` extension method for Product to allow product search list and paginated list selector to continue working with the full objects. This is to keep the scope of this PR small.
- Make `createStockText` a String extension to be reusable across different types, not just `ProductFormDataModel`.

## Testing steps
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

- Log in to a test store.
- Navigate to the Products tab and confirm that the product lis loaded correctly.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested and confirmed on simulator iPhone 16 iOS 18.4.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

No UI changes.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add missing properties

itsmeichigo committed:
Make some entities conform to ListItemConvertible with default implementation

itsmeichigo committed:
Replace full product objects in product list

itsmeichigo committed:
Fix swiftlint issue

itsmeichigo committed:
Merge branch 'woomob-619-xcode-warnings-performing-io-on-the-main-thread-can-cause' into woomob-619-product-list-update

itsmeichigo committed:
Move conversion from full object to separate file

itsmeichigo committed:
Update tests

itsmeichigo committed:
Merge branch 'woomob-619-xcode-warnings-performing-io-on-the-main-thread-can-cause' into woomob-619-product-list-update

itsmeichigo committed:
Unify logic for createStockText

itsmeichigo committed:
Revert changes made to PaginatedListSelectorViewController

itsmeichigo committed:
Revert changes made to SearchUICommand and related files

itsmeichigo committed:
Revert redundant changes

itsmeichigo committed:
Remove redundant conformance

itsmeichigo committed:
Update release notes

RafaelKayumov commented on Modules/Sources/Yosemite/Tools/ResultsController.swift:
```diff
@@ -357,4 +357,8 @@ public extension ResultsController where T: ListItemConvertible {
+
+    func listItem(at indexPath: IndexPath) -> T.ListItemType {
+        return controller.object(at: indexPath).toListItem()
+    }
```
👍

RafaelKayumov reviewed (approved):
LGTM. Products, variations, bundles look fine.

itsmeichigo committed:
Merge branch 'woomob-619-xcode-warnings-performing-io-on-the-main-thread-can-cause' into woomob-619-product-list-update

itsmeichigo committed:
Add ProductListItem on the view layer

itsmeichigo committed:
Merge branch 'woomob-619-xcode-warnings-performing-io-on-the-main-thread-can-cause' into woomob-619-product-list-update

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/252">23.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Remove added tests for ResultsController

itsmeichigo committed:
Merge branch 'woomob-619-xcode-warnings-performing-io-on-the-main-thread-can-cause' into woomob-619-product-list-update

itsmeichigo committed:
Add workaround for using simplified objects for display in list items

itsmeichigo committed:
Fix swiftlint

itsmeichigo commented:
@RafaelKayumov I also added changes in this PR to use a separate type for product list on the view layer. Due to complications with batch product update, I added a workaround to get list items for the product list without using the generic results controller as discussed in #15959. Please take another look when you can.

cc @jaclync

itsmeichigo committed:
Merge branch 'trunk' into woomob-619-product-list-update

itsmeichigo committed:
Remove unused properties

itsmeichigo committed:
Merge branch 'trunk' into woomob-619-product-list-update

itsmeichigo committed:
Ignore periphery for OrderDetailsProduct initializer

itsmeichigo committed:
Restore initializer for ProductDetailsCellViewModel

itsmeichigo committed:
Remove unused property
